### PR TITLE
Added RandomizedSearchCV functionality, with tests.

### PR DESCRIPTION
**Pull request by gnsiva**
Addresses: https://github.com/databricks/spark-sklearn/issues/15

I've added a spark distributed version of `RandomizedSearchCV`.

To do this, I've moved the `_fit()` method that was added for `GridSearchCV` to a base class and then used that as the super class for both `GridSearchCV` and `RandomizedSearchCv`.

### CODE REVIEW

**Comment by gnsiva** [comment]
Hi, is there a problem with this merge request?

**Comment by wallejak** [comment]
Hi there, will the RandomizedSearchCV be added?

**Comment by akurniawan** [comment]
this is an amazing PR! any ETA for when this PR will be merged into the master?

**Comment by gnsiva** [comment]
@thunterdb @smurching if you approve the method used I will update the PR to fix the conflicts which have now built up.

File: python/spark_sklearn/grid_search.py [hunk 1]
```diff
@@ -180,108 +295,184 @@ def fit(self, X, y=None):
...
+        self.sc = sc
+        self.param_distributions = param_distributions
+        self.n_iter = n_iter
+        self.random_state = random_state
+        super(RandomizedSearchCV, self).__init__(
+            estimator=estimator, scoring=scoring, fit_params=fit_params,
+            n_jobs=n_jobs, iid=iid, refit=refit, cv=cv, verbose=verbose,
+            pre_dispatch=pre_dispatch, error_score=error_score)
+
+    def fit(self, X, y=None):
```
Comment (WeichenXu123) [inline python/spark_sklearn/grid_search.py]: Why you remove the `group` parameter of `fit` ?

File: python/spark_sklearn/grid_search.py [hunk 2]
```diff
@@ -10,8 +10,123 @@
...
+        if self.refit:
+            # fit the best estimator using the entire dataset
+            # clone first to work around broken estimators
+            best_estimator = clone(base_estimator).set_params(
+                **best.parameters)
+            if y is not None:
+                best_estimator.fit(X, y, **self.fit_params)
+            else:
+                best_estimator.fit(X, **self.fit_params)
+            self.best_estimator_ = best_estimator
```
Comment (WeichenXu123) [inline python/spark_sklearn/grid_search.py]: Where is `cv_results_` exists in current master version ?
It seems that you change a lot here.

File: python/spark_sklearn/grid_search.py (hunk 1 above)
Comment (gnsiva) [reply python/spark_sklearn/grid_search.py]: I didn't, when I submitted the PR a year ago that parameter wasn't there.

**Comment by gnsiva** [comment]
Thanks for the interest, I will update the PR today hopefully

**Comment by gnsiva** [comment]
@WeichenXu123 sorry for the delay, I've updated the pull request. 
The change is quite simple, the only difference between {RandomizedSearchCV} and {GridSearchCV} are the init methods and the fit method (which calls _fit()). This is the same as in scikit-learn.

I have then abstracted out the spark_sklearn logic from {GridSearchCV} to {SparkBaseSearchCV}.

**Comment by aadu** [comment]
Any update on this? RandomizedSearchCV is pretty critical for me so going to use @gnsiva's fork until this get's merged.

**Comment by srowen** [comment]
First, @gnsiva this is way way overdue, sorry about that. I rebased this and added some changes in https://github.com/databricks/spark-sklearn/pull/100 . If you're interested in reviewing, have a look. Credit to you for implementing this a long time ago, and we should be able to release this soon.

**Comment by srowen** [comment]
Thanks again @gnsiva ; I merged #100

**Pull request closed without merging**